camera/rpi-files/data_archiver.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+
+"""Data archiving module.
+
+DataArchiver saves batches in a selected format.
+"""
+
+import csv
+import os
+from abc import ABC
+from enum import Enum
+
+from data_batcher import Batch
+
+
+class EventType(Enum):
+    """Enum for event types."""
+
+    monitoring_started = 0
+    monitoring_ended = 1
+    person_entered = 2
+    person_left = 3
+
+
+class DataArchiver(ABC):
+    """Archives batches of data."""
+
+    def init(self):
+        """Initialize the archive if not initialized."""
+
+    def append_event(self, timestamp: float, event_type: Enum):
+        """Includes single event in the archive. May not flush archive yet.
+
+        Args:
+            timestamp: event timestamp
+            event_type: event type
+        """
+
+    def append(self, batch: Batch):
+        """Include batch in the archive. May not flush archive yet.
+
+        Args:
+            batch: new data batch to include
+        """
+
+    def flush(self):
+        """Force flushing the archive."""
+
+    def finalize(self):
+        """Finalize the archive. No more data can be saved."""
+
+
+class CsvArchiver(DataArchiver):
+    """Archive batches of data to a file."""
+
+    def __init__(self, path):
+        """Save path for writing.
+
+        Raises:
+            TypeError: if path is not a string
+
+        Args:
+            path: file path
+        """
+        if not isinstance(path, str):
+            msg = 'Path `{0}` is not a string'.format(path)
+            raise TypeError(msg)
+        self._path = path
+        self._archive_file = None
+        self._writer = None
+        self._entries = {}
+
+    def init(self, force_overwrite=False):
+        """Create file for writing.
+
+        Args:
+            force_overwrite: force overwriting file
+
+        Raises:
+            RuntimeError: if path already exists or could not create file
+        """
+        if not force_overwrite and os.path.exists(self._path):
+            msg = 'Path `{0}` already exists'.format(self._path)
+            raise RuntimeError(msg)
+
+        try:
+            self._archive_file = open(self._path, 'w+')  # noqa: WPS515
+        except OSError as ex1:
+            msg = 'Could not open path `{0}` for writing: {1}'.format(
+                self._path, ex1,
+            )
+            raise RuntimeError(msg)
+
+        field_names = [
+            'timestamp',
+            EventType.monitoring_started.name,
+            EventType.monitoring_ended.name,
+            EventType.person_entered.name,
+            EventType.person_left.name,
+        ]
+        try:
+            self._writer = csv.DictWriter(
+                self._archive_file, fieldnames=field_names,
+            )
+        except csv.Error as ex2:
+            msg = 'Error creating csv.writer: {0}'.format(ex2)
+            raise RuntimeError(msg)
```

Comment:
How to you plan to handle these Runtime errors? Is it gonna try to restart / reboot or just hang after sending some signal

Reply:
I think we can close the device. All the exceptions are handled, but if camera can't do it's work, I wouldn't try to restart.